crates/node-bindings/src/parcel/parcel.rs

```diff
@@ -25,30 +37,86 @@ impl ParcelNapi {
 
+    // Parcel Core Options
+    let mut parcel_options = ParcelOptions::default();
+
     // Wrap the JavaScript-supplied FileSystem
-    let fs = FileSystemNapi::from_options(&env, &options)?;
+    parcel_options.fs = FileSystemNapi::from_options(&env, &options)?;
+
+    // Assign Rust thread count from JavaScript
+    let js_threads = options.get_named_property::<JsUnknown>("threads")?;
+    parcel_options.threads = match js_threads.get_type()? {
+      napi::ValueType::Undefined => Ok(parcel_options.threads),
+      napi::ValueType::Number => Ok(js_threads.coerce_to_number()?.get_uint32()? as usize),
+      _ => Err(napi::Error::from_reason("Expected number for threads")),
+    }?;
 
     // Set up Nodejs plugin bindings
-    let node_workers: JsNumber = options.get_property(env.create_string("nodeWorkers")?)?;
-    let node_workers = node_workers.get_uint32()?;
-    let rpc_host_nodejs =
-      RpcHostNodejs::new(&env, options.get_named_property("rpc")?, node_workers)?;
-
-    // Initialize Parcel
-    let parcel = Parcel::new(ParcelOptions {
-      fs,
-      rpc: Some(Arc::new(rpc_host_nodejs)),
-    });
+    let js_node_workers = options.get_named_property::<JsUnknown>("nodeWorkers")?;
+    let node_worker_count = match js_node_workers.get_type()? {
+      napi::ValueType::Undefined => Ok(parcel_options.threads.clone()),
+      napi::ValueType::Number => Ok(js_node_workers.coerce_to_number()?.get_uint32()? as usize),
+      _ => Err(napi::Error::from_reason("Expected number for threads")),
+    }?;
 
+    let rpc_host_nodejs = RpcHostNodejs::new(
+      &env,
+      options.get_named_property("rpc")?,
+      node_worker_count.clone(),
+    )?;
+    parcel_options.rpc = Some(Arc::new(rpc_host_nodejs));
+
+    // Return self
     Ok(Self {
-      parcel: Arc::new(parcel),
+      node_worker_count: node_worker_count as u32,
+      parcel: Arc::new(Parcel::new(parcel_options)),
     })
   }
 
   #[napi]
-  pub async fn build(&self) -> napi::Result<()> {
-    self.parcel.build().unwrap();
-    Ok(())
+  pub fn build(
+    &self,
+    env: Env,
+    _options: JsObject,
+    callback: JsFunction,
+  ) -> napi::Result<JsUndefined> {
+    let (tx, rx) = channel();
+    // Parse build options from JS options
+    let build_options = BuildOptions {};
+
+    // Call build in its own dedicated system thread
+    thread::spawn({
+      let parcel = self.parcel.clone();
+      move || tx.send(parcel.build(build_options))
+    });
+
+    // JavaScript bindings to handle the build result
+    let callback: ThreadsafeFunction<anyhow::Result<BuildResult>> = env
+      .create_threadsafe_function(
+        &callback,
+        0,
+        |ctx: ThreadSafeCallContext<anyhow::Result<BuildResult>>| match ctx.value {
+          Ok(_) => Ok(vec![ctx.env.create_object()]),
+          Err(err) => Err(napi::Error::from_reason(format!("{:?}", err))),
+        },
+      )?;
+
+    // Asynchronously wait for a response from the build
+    env.spawn_future(async move {
```

Comment:
I don't think you need tokio or callbacks for this. You can use `env.create_deferred` to create a promise, and return that.

https://github.com/parcel-bundler/parcel/blob/c03a29ede215967a9ec4e472a1161eed0a913af6/crates/node-bindings/src/core.rs#L128-L137

Reply:
Nice, will use the deferred API!

Reply:
Updated